src/routing/stage/approved.rs

```diff
@@ -1842,70 +1887,69 @@ impl Approved {
             return Ok(vec![]);
         };
 
-        let hash = msg.hash();
-        let actions = self.shared_state.update_section_knowledge(
-            &self.node_info.name(),
-            src_prefix,
-            src_key,
-            msg.dst_key().as_ref(),
-            hash,
-        );
+        let is_neighbour = self.section.prefix().is_neighbour(src_prefix);
 
-        let mut commands = vec![];
+        let mut commands = Vec::new();
+        let mut vote_send_neighbour_info = false;
 
-        for action in actions {
-            match action {
-                VoteTheirKey { prefix, key } => {
-                    commands.extend(self.vote(Vote::TheirKey { prefix, key })?);
-                }
-                VoteTheirKnowledge { prefix, key_index } => {
-                    commands.extend(self.vote(Vote::TheirKnowledge { prefix, key_index })?);
-                }
-                SendNeighbourInfo { dst, nonce } => commands.extend(self.send_neighbour_info(
-                    dst,
-                    nonce,
-                    self.shared_state.sections.key_by_name(&dst.name()).cloned(),
-                )?),
+        if !src_prefix.matches(&self.node.name()) && !self.network.has_key(src_key) {
+            // Only vote `TheirKeyInfo` for non-neighbours. For neighbours, we update the keys
+            // via `NeighbourInfo`.
+            if is_neighbour {
+                vote_send_neighbour_info = true;
+            } else {
+                commands.extend(self.vote(Vote::TheirKey {
+                    prefix: *src_prefix,
+                    key: *src_key,
+                })?);
+            }
+        }
+
+        if let Some(dst_key) = msg.dst_key() {
+            let old = self.network.knowledge_by_section(src_prefix);
+            let new = self.section.chain().index_of(dst_key).unwrap_or(0);
```

Comment:
I am bit confused here.
This `new` is the src knowledge of us, and the `old` is our knowledge of src
but following we are comparing these two separated knowledge ?

Reply:
No, `old` is their knowledge of our section that we currently know of and `new` is their knowledge of our section according to the message received. We simply check whether `new` is newer than `old` and update it if that's the case.